### DISCUSSION TITLE
How to make a subclass that can take a modified FEProblemSolve::solve() into it?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#27695] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi guys,

so I have modified FEProblemSolve::solve() in MOOSE folder (/moose_beh/moose/framework/src/executioners/FEProblemSolve.C)
such that it moves on to the next step even if the simulation step isn't converged

```
bool
FEProblemSolve::solve()
{
  // This loop is for nonlinear multigrids (developed by Alex)
  for (MooseIndex(_num_grid_steps) grid_step = 0; grid_step <= _num_grid_steps; ++grid_step)
  {
    _problem.solve(_nl.number());

    if (_problem.shouldSolve())
    {
      if (_problem.converged(_nl.number()))
        _console << COLOR_GREEN << " Solve Converged!" << COLOR_DEFAULT << std::endl;
      else
      {
        _console << COLOR_RED << " Solve Did NOT Converge!" << COLOR_DEFAULT << std::endl;
        // return false;
      }
    }
    else
      _console << COLOR_GREEN << " Solve Skipped!" << COLOR_DEFAULT << std::endl;

    if (grid_step != _num_grid_steps)
      _problem.uniformRefine();
  }
  // return _problem.converged(_nl.number());
  return true;
}
```

but I don't want this to be part of the main MOOSE code, I want MOOSE code to remain clean, and do everything in my own app.

I want to make a subclass that does exactly the same thing as, let's call it 'TransientMod'. How do I bring this modified `FEProblemSolve::solve()` into it?
I checked doxegen and i dont see the connection between  Transient.C and FEProblemSolve.C at all.
However, they are in the same folder (src/executioners) in MOOSE main src.

Kind regards,
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Hello

You can either make a new problem class and specify in the s problem block to use it

Or you can wait a month or so and we will introduce customized convergence objects. These will let you do what you want. The decision on whether something is converged or not will no longer belong to the problem but to a convergence system, of which you will be able to make custom classes 

Guillaume

**Reply by Traiwit** [reply]
wow `customized convergence objects` sounds awesome to me! I can live with this for now.

Thank you @GiudGiud!

Kind regards,
Trai

**Reply by YaqiWang** [reply]
This line in `Transient.C` sets up `_fixed_point_solve->setInnerSolve(_feproblem_solve);`. You can build an object with your modified `FEProblemSolve` in you `TransientMod` and set it as the inner object of the existing `_fixed_point_solve`.
